include/sisl/fds/buffer.hpp

```diff
@@ -300,24 +301,48 @@ struct io_blob : public blob {
 /* An extension to blob where the buffer it holds is allocated by constructor and freed during destruction. The only
  * reason why we have this instead of using vector< uint8_t > is that this supports allocating in aligned memory
  */
-struct byte_array_impl : public io_blob {
-    byte_array_impl(const uint32_t sz, const uint32_t alignment = 0, const buftag tag = buftag::common) :
-            io_blob(sz, alignment, tag),
-            m_tag{tag} {}
-    byte_array_impl(uint8_t* const bytes, const uint32_t size, const bool is_aligned) :
-            io_blob(bytes, size, is_aligned) {}
-    ~byte_array_impl() { io_blob::buf_free(m_tag); }
-
+struct io_blob_safe : public io_blob {
+public:
     buftag m_tag;
+
+public:
+    io_blob_safe(uint32_t sz, uint32_t alignment = 0, buftag tag = buftag::common) :
+            io_blob(sz, alignment, tag), m_tag{tag} {}
+    io_blob_safe(uint8_t* bytes, uint32_t size, bool is_aligned) : io_blob(bytes, size, is_aligned) {}
+    virtual ~io_blob_safe() {
+        if (bytes != nullptr) { io_blob::buf_free(m_tag); }
+    }
+
+    io_blob_safe(io_blob_safe const& other) = delete;
+    io_blob_safe(io_blob_safe&& other) : io_blob{other.bytes, other.size, other.is_aligned} {
+        other.bytes = nullptr;
+        other.size = 0;
+    }
+
+    io_blob_safe& operator=(io_blob_safe const& other) = delete; // Delete copy constructor
+    io_blob_safe& operator=(io_blob_safe&& other) {
+        if (bytes != nullptr) { this->buf_free(m_tag); }
```

Comment:
Why not swap and let the caller's scope take care of it?

Reply:
Can we technically swap, given the semantics of std::move() is that post move, the moved entity should be reset? I am worried, if someone uses the value after move, would give indifferent result

```
io_blob_safe b1{ptr1, 1024};
io_blob_safe b2{ptr2, 10};
auto b2 = std::move(b);
...
...
std::cout << b.size << "\n" // Shouldn't this required to be report 0 and b.bytes have nullptr?
```

Reply:
No, i think it's meant to be indeterministic though.